### ISSUE TITLE
Chatterbot data to 2M is very slow to reflect why?

### ISSUE THREAD

**Issue by yeshoujquery** [gunthercox/ChatterBot#679]
4G memory usage is not particularly large, but after the training data reflects a very slow data only 2M did not use mongodb to answer

**Comment by gunthercox** [comment]
If you are using the default JSON file storage adapter then the **entire** file has to be loaded into memory to search for matching statements. It works well for testing and very small data sets but it is completely impractical for larger data sets. This is the main reason that a real database is highly recommended.

**Comment by yeshoujquery** [comment]
@gunthercox Can you give me a way to write it? If I use the database in the case of trouble to help me stick to your implementation code I learn about my learning chatterbot for a long time has been reflected very slowly on the very distress

**Comment by gunthercox** [comment]
Sure, to clarify what I meant, by default the chat bot will use the value `chatterbot.storage.JsonFileStorageAdapter` which is a file based database. The file database can become very slow when large amounts of data are added.

You can switch the parameter to make is use a database such as Mongo DB, which has much better performance.

Here is an example of what this might look like in your code.

```python
chatbot = ChatBot(
    "My ChatterBot",
    storage_adapter="chatterbot.storage.MongoDatabaseAdapter"
)
```

There is also more details on each of the storage adapters in ChatterBot's documentation if you are interested in reading about it http://chatterbot.readthedocs.io/en/latest/storage/index.html

**Comment by yeshoujquery** [comment]
@gunthercox
 I used the mongodb training 373753 data query once the time is 315831ms what way can be faster? This speed really a bit unacceptable ... seeking answers

`chatbotEnglish = ChatBot(
    "english mongodb",
    storage_adapter='chatterbot.storage.MongoDatabaseAdapter',
    logic_adapters=[
        "chatterbot.logic.MathematicalEvaluation",
        "chatterbot.logic.TimeLogicAdapter",
        'chatterbot.logic.BestMatch'
    ],
    preprocessors = [
        'chatterbot.preprocessors.clean_whitespace',
        'chatterbot.preprocessors.unescape_html'
    ],
    filters=[
        'chatterbot.filters.RepetitiveResponseFilter'
    ],
    input_adapter="chatterbot.input.VariableInputTypeAdapter",
    output_adapter="chatterbot.output.OutputAdapter",
    trainer='chatterbot.trainers.ChatterBotCorpusTrainer',
    database='chatterbot-database',
    read_only=False
)`

**Comment by yeshoujquery** [comment]
`        
  for statement in statement_list:
            confidence = self.compare_statements(input_statement, statement)

            if confidence > closest_match.confidence:
                statement.confidence = confidence
                closest_match = statement
`

There has been a cycle in which all cause special slow query data fairly fast here loop matching accuracy can be maintained by the author?

**Comment by yeshoujquery** [comment]
@gunthercox 
`
for statement in statement_list:
       confidence = self.compare_statements(input_statement, statement)
        if confidence > closest_match.confidence:
            statement.confidence = confidence
            closest_match = statement
`

There has been a cycle in which all cause special slow query data fairly fast here loop matching accuracy can be maintained by the author?

**Comment by whm2300** [comment]
@yeshoujquery. I have the same problem,  do you fix it?

**Comment by yeshoujquery** [comment]
@whm2300 Did not solve the problem, the author seems to be unable to solve the problem should be the framework itself, I look at the conclusion of the source code

**Comment by whm2300** [comment]
thanks 

发自我的 iPhone

**Comment by vkosuri** [comment]
@yeshoujquery I'll try to reproduce this problem on my machine next week, will get it back soon.

**Comment by yeshoujquery** [comment]
@vkosuri That's better, waiting for your good news, can you let me know?

**Comment by vkosuri** [comment]
sure

**Comment by yeshoujquery** [comment]
@vkosuri Thanks very much

**Comment by yeshoujquery** [comment]
@vkosuri Upload it, or can you give me a copy?

**Comment by yeshoujquery** [comment]
@vkosuri Where is the code chatterbot seems to be particularly slow to solve

**Comment by yeshoujquery** [comment]
@gunthercox  This problem can be solved using the mongodb also useless, or very slow

**Comment by anthdm** [comment]
The problem is rather simple to detect. Comparing more than 2000 responses with lexical comparison makes it super slow. Nothing to do with MongoDB. 

A solution can be:
compute the word vectors for each statement to determine the vector for the whole phrase.
Rinse and repeat for the user input. The only thing we then having to do is query the vectors in MongoDB on closest similarity (Cosine Similarity). This will also be more accurate than lexical comparison.

Drawback:
Loading in a good and accurate word embeddings model into memory will take time.

Some tips on where to start Google for word2vec, doc2vec, glove, fast text.

**Comment by yeshoujquery** [comment]
@vkosuri Mongodb 373753 data time-consuming 89ms, relatively fast before, but can it be faster? Seconds out of that kind?

**Comment by yeshoujquery** [comment]
@anthdm You say the solution, I don't choose chatterbot, then I directly choose tensorflow, now want to use chatterbot

**Comment by vkosuri** [comment]
@yeshoujquery is it possible share your train data? I will debug further more on your train data

**Comment by yeshoujquery** [comment]
@vkosuri Some Chinese data, in fact, as long as the order of magnitude will be slow,Do you need? Or chatterbot, how many orders of maximum support?

**Comment by vkosuri** [comment]
@yeshoujquery I took [English corpus](https://github.com/gunthercox/chatterbot-corpus/tree/master/chatterbot_corpus/data/english), made it as 8MB data, started debugging will update more on this issue.

**Comment by yeshoujquery** [comment]
@vkosuri  Well, thank you very much. The amount of data will be slow, and no way out

**Comment by yeshoujquery** [comment]
@vkosuri Thank you very much. I'm waiting for your good news @gunthercox

**Comment by vkosuri** [comment]
Trained 8.9MB data, with some minor changes observed fallowing stats on machine. i7, 8GB

<pre>
(venv)vkosuri@vkosuri:~/ChatterBot$ python terminal_mongo_example.py 
Started Trainig
Trainig completed
<b>148.540009022 seconds</b>
2017-05-15,17:59:09.073 INFO {input_adapter} [process_input_statement] Recieved input statement: I enjoy raspberry ice cream.
2017-05-15,17:59:09.074 INFO {input_adapter} [process_input_statement] "I enjoy raspberry ice cream." is a known statement
2017-05-15,17:59:09.180 INFO {best_match} [process] Using "I enjoy raspberry ice cream." as a close match to "Is it cramped in the computer"
2017-05-15,17:59:09.182 INFO {best_match} [process] Selecting response from 1 optimal responses.
2017-05-15,17:59:09.182 INFO {response_selection} [get_first_response] Selecting first response from list of 1 options.
2017-05-15,17:59:09.182 INFO {best_match} [process] Response selected. Using "No."
2017-05-15,17:59:09.183 INFO {multi_adapter} [process] BestMatch selected "No." as a response with a confidence of 0.46
2017-05-15,17:59:09.183 INFO {multi_adapter} [process] NoKnowledgeAdapter selected "I enjoy raspberry ice cream." as a response with a confidence of 0
No.
<b>0.111059904099 seconds</b>
2017-05-15,17:59:09.184 INFO {input_adapter} [process_input_statement] Recieved input statement: I enjoy raspberry ice cream.
2017-05-15,17:59:09.185 INFO {input_adapter} [process_input_statement] "I enjoy raspberry ice cream." is a known statement
2017-05-15,17:59:09.393 INFO {best_match} [process] Using "I enjoy raspberry ice cream." as a close match to "I enjoy raspberry ice cream."
2017-05-15,17:59:09.396 INFO {best_match} [process] No response to "I enjoy raspberry ice cream." found. Selecting a random response.
2017-05-15,17:59:09.397 INFO {multi_adapter} [process] BestMatch selected "Do you feel." as a response with a confidence of 0
2017-05-15,17:59:09.397 INFO {multi_adapter} [process] NoKnowledgeAdapter selected "I enjoy raspberry ice cream." as a response with a confidence of 0
2017-05-15,17:59:09.397 INFO {chatterbot} [learn_response] Adding "I enjoy raspberry ice cream." as a response to "No."
Do you feel.
<b>0.21406006813 seconds</b>
</pre>

**Comment by vkosuri** [comment]
@yeshoujquery Did you observed any improvement with new changes?

**Comment by chasingML** [comment]
@yeshoujquery hi. u used  373753 data with mongodb  time-consuming 89ms? right? I want to know how to make it ? I used 7w data with mongodb  time-consuming 41s for response, it's really too slow? I do not clear what's wrong with my codes? tks a lot!!!

**Comment by yeshoujquery** [comment]
@gunthercox @vkosuri Is the speed up now? Compared to previous versions

**Comment by yeshoujquery** [comment]
@chasingML Do you try a new version or is it slow? Or give me the data for a try

**Comment by vkosuri** [comment]
@yeshoujquery Still i haven't tested

**Comment by yeshoujquery** [comment]
@vkosuri The version is upgraded to 0.8+, and the speed is still not guaranteed？

**Comment by yeshoujquery** [comment]
@vkosuri If speed is not guaranteed to do more work, it is not done.

**Comment by zxsimple** [comment]
@gunthercox @yeshoujquery 

how you improved the performance?
I trained with Baidu WebQA data http://paddlepaddle.bj.bcebos.com/dataset/webqa/WebQA.v1.0.zip , only for 3000 testing data.

It takes around 3s for single response, I ran it on my laptop with MongdbAdapter.

Also @mymusise  https://github.com/gunthercox/ChatterBot/pull/945 mentioned mongdb inverted index can improve the performance a lot.

What's the plan or solution for the mongodb performance?

**Comment by zxsimple** [comment]
`        for statement in statement_list:
            confidence = self.compare_statements(input_statement, statement)
            if confidence > closest_match.confidence:
                statement.confidence = confidence
                closest_match = statement`

The for loop to get the closest statement takes O(n) time, nothing related to MongoDB. 3K question data takes around 3s. 

@vkosuri how you made it to be 89ms for 373753 recodes?

**Comment by vkosuri** [comment]
By using list comprehension on different places https://github.com/gunthercox/ChatterBot/pull/738/files and https://github.com/gunthercox/ChatterBot/pull/823/files

**Comment by zxsimple** [comment]
@vkosuri I tried your branch vkosuri/ChatterBot-listcomprehension, it improved from 3s to 1.9s, not so fast as you did.

**Comment by zxsimple** [comment]
@yeshoujquery would you please try this: [https://github.com/gunthercox/ChatterBot/pull/1158](https://github.com/gunthercox/ChatterBot/pull/1158)

**Comment by yeshoujquery** [comment]
@zxsimple Have you tried very well? I can't find the data.

**Comment by zxsimple** [comment]
@yeshoujquery, yes, the response improved from 3s to 190ms after the change. You can install python-Levenshtein package and see whether it has better performance. If so #1158 can be helpful.
